cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -346,14 +374,396 @@ void AddSortingKernels(VectorKernel base, VectorFunction* func) {
+        status_ = sort_keys_[i].type->Accept(this);
+        if (current_compared_ != 0) {
+          break;
+        }
+      }
+      return current_compared_ < 0;
+    }
+
+#define VISIT(TYPE)                                \
+  Status Visit(const TYPE##Type& type) override {  \
+    current_compared_ = CompareType<TYPE##Type>(); \
+    return Status::OK();                           \
+  }
+
+    VISIT(Int8)
+    VISIT(Int16)
+    VISIT(Int32)
+    VISIT(Int64)
+    VISIT(UInt8)
+    VISIT(UInt16)
+    VISIT(UInt32)
+    VISIT(UInt64)
+    VISIT(Float)
+    VISIT(Double)
+    VISIT(String)
+    VISIT(Binary)
+    VISIT(LargeString)
+    VISIT(LargeBinary)
+
+#undef VISIT
+
+   private:
+    template <typename Type>
+    int32_t CompareType() {
+      using ArrayType = typename TypeTraits<Type>::ArrayType;
+      const auto& sort_key = sort_keys_[current_sort_key_index_];
+      auto order = sort_key.order;
+      int64_t index_left = 0;
+      auto array_left = sort_key.ResolveChunk<ArrayType>(current_left_, index_left);
+      int64_t index_right = 0;
+      auto array_right = sort_key.ResolveChunk<ArrayType>(current_right_, index_right);
+      if (sort_key.null_count > 0) {
+        auto is_null_left = array_left->IsNull(index_left);
+        auto is_null_right = array_right->IsNull(index_right);
+        if (is_null_left && is_null_right) {
+          return 0;
+        } else if (is_null_left) {
+          return 1;
+        } else if (is_null_right) {
+          return -1;
+        }
+      }
+      auto left = array_left->GetView(index_left);
+      auto right = array_right->GetView(index_right);
+      int32_t compared;
+      if (left == right) {
+        compared = 0;
+      } else if (left > right) {
+        compared = 1;
+      } else {
+        compared = -1;
+      }
+      if (order == SortOrder::DESCENDING) {
+        compared = -compared;
+      }
+      return compared;
+    }
+
+    Status status_;
+    std::vector<ResolvedSortKey> sort_keys_;
+    int64_t current_left_;
+    int64_t current_right_;
+    size_t current_sort_key_index_;
+    int32_t current_compared_;
+  };
+
+ public:
+  TableSorter(uint64_t* indices_begin, uint64_t* indices_end, const Table& table,
+              const SortOptions& options)
+      : indices_begin_(indices_begin),
+        indices_end_(indices_end),
+        comparer_(table, options.sort_keys) {}
+
+  Status Sort() {
+    ARROW_RETURN_NOT_OK(comparer_.status());
+    return comparer_.sort_keys()[0].type->Accept(this);
+  }
+
+#define VISIT(TYPE) \
+  Status Visit(const TYPE##Type& type) override { return SortInternal<TYPE##Type>(); }
+
+  VISIT(Int8)
+  VISIT(Int16)
+  VISIT(Int32)
+  VISIT(Int64)
+  VISIT(UInt8)
+  VISIT(UInt16)
+  VISIT(UInt32)
+  VISIT(UInt64)
+  VISIT(Float)
+  VISIT(Double)
+  VISIT(String)
+  VISIT(Binary)
+  VISIT(LargeString)
+  VISIT(LargeBinary)
+
+#undef VISIT
+
+ private:
+  template <typename Type>
+  Status SortInternal() {
```

Comment:
Can you add comments when you define a non-trivial class or method?

Reply:
Added.